### ISSUE TITLE
sax_exception_example does not compile under mingw

### ISSUE THREAD

**Issue by neheb** [libxmlplusplus/libxmlplusplus#45]
```
FAILED: subprojects/libxml++-5.0.2/examples/sax_exception_example.exe
x86_64-w64-mingw32-g++  -o subprojects/libxml++-5.0.2/examples/sax_exception_example.exe subprojects/libxml++-5.0.2/examples/sax_exception_example.exe.p/sax_exception_main.cc.obj subprojects/libxml++-5.0.2/examples/sax_exception_example.exe.p/sax_exception_myparser.cc.obj -Wl,--allow-shlib-undefined -Wl,--start-group subprojects/libxml++-5.0.2/libxml++/libxml++-5.0.dll.a /usr/x86_64-w64-mingw32/sys-root/mingw/lib/libxml2.dll.a -Wl,--subsystem,console -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -Wl,--end-group
/usr/lib/gcc/x86_64-w64-mingw32/12.2.1/../../../../x86_64-w64-mingw32/bin/ld: subprojects/libxml++-5.0.2/examples/sax_exception_example.exe.p/sax_exception_myparser.cc.obj:myparser.cc:(.rdata$.refptr._ZTVN5xmlpp9exceptionE[.refptr._ZTVN5xmlpp9exceptionE]+0x0): undefined reference to `vtable for xmlpp::exception'
collect2: error: ld returned 1 exit status
```

**Comment by kjellahl** [comment]
It looks like the program has been compiled, but can't be linked. The message
*undefined reference to 'vtable for xmlpp::exception'* comes from the linker.

Can you compile and link other example programs? examples/sax_parser and
examples/dom_parse_entities are interesting. They also throw exceptions of
type xmlpp::exception or one of its subclasses.

I don't use mingw. I would be fine if we can solve this issue without me
installing and using it. I've never used libxml++ on Windows.

**Comment by neheb** [comment]
The others get linked properly.

As far as MinGW goes, you can cross compile in Linux. Most distributions come with a mingw toolchain.

**Comment by kjellahl** [comment]
I've installed mingw-w64 on my Ubuntu/Linux system, and cross compiled.
I get the same error (undefined reference to vtable for xmlpp::exception).

@fanc999 mingw can cross compile (from Linux to Windows) and build everything,
including examples/sax_exception, if I modify libxml++/exceptions/exception.h.
Remove all LIBXMLPP_API from within the xmlpp::exception class and add it on
```
class LIBXMLPP_API exception : public std::exception
```

Other observations when using mingw on Linux:

`xmlxx_res = windows.compile_resources(xmlxx_rc)` in libxml++/meson.build failed.
There is no resource compiler. Perhaps I can install one, but is it necessary?
I changed `if host_machine.system() == 'windows'` to `if is_msvc`.

I get compiler warnings from `friend` declarations:
```
../libxml++/document.h:75:29: warning: attribute ignored in declaration of ‘class xmlpp::SaxParser’ [-Wattributes]
   75 |   friend LIBXMLPP_API class SaxParser;
      |                             ^~~~~~~~~
../libxml++/document.h:75:29: note: attribute for ‘class xmlpp::SaxParser’ must follow the ‘class’ keyword
```
I changed `friend LIBXMLPP_API class` to `friend class LIBXMLPP_API` in 3 header files.
Everything then builds without warnings.

I have not committed any of these modifications. I hope I can get some comments.

I have not tried to run the example programs. It would require an exe_wrapper
that I haven't installed.

**Comment by neheb** [comment]
Right. wine64 typically.

The resource compiler is packaged at least on fedora. No idea about Ubuntu. MSYS2 also includes it.

edit: the LIBXMLPP_API changes fix compilation.

**Comment by fanc999** [comment]
Hi @kjellahl,

Sadly, there is a reason I had the `LIBXMLPP_API` in the individual members of `xmlpp::exception` instead of over the entire class because there is an `ustring message_` private member, which would cause warning C4251, which would lead to issues like [1].  I'll see what I can do here, since I don't have a MinGW/MSYS environment...

The resource compiler is to embed the package version info into the built libxml++ DLL, which I think can be found in a package like doing `sudo apt-get install binutils-mingw-w64-i686` or `sudo apt-get install binutils-mingw-w64-x86_64` on an Ubuntu system (I think).  It probably needs to be defined in your Meson cross-compilation file (if it's too cumbersome for you to install it, I think you can just disable it momentarily, but I do think that it should be used normally, since it's used by quite a number of ofther projects).

For the `friend`'s, the `LIBXMLPP_API` must be where they are, otherwise the build breaks on Visual Studio.  I think we could get away dropping the `LIBXMLPP_API` from the `friend`'s, at least on Visual Studio.

It turns out that we might need to add a new macro <strike>try defining `LIBXMLPP_API`</strike> for such classes for GCC (Windows) builds by <strike>to be also nothing</strike> when building `libxml++` as a DLL, or as `__declspec(dllexport) __attribute__((visibility("default")))`, i.e, in `libxml++config.h.[in|meson]:

```
#ifdef LIBXMLPP_DLL
  #ifdef LIBXMLPP_BUILD
    #define LIBXMLPP_API __declspec(dllexport)
    #ifdef __GNUC__
      #define LIBXMLPP_VISIBILITY_DEFAULT __declspec(dllexport) __attribute__((visibility("default")))
      #define LIBXMLPP_MEMBER_METHOD
    #endif
  #elif !defined (__GNUC__)
    #define LIBXMLPP_API __declspec(dllimport)
  #else /* don't dllimport classes/methods on GCC/MinGW */
    #define LIBXMLPP_API
  #endif /* LIBXMLPP_BUILD - DLL */
#else
  /* Build a static library or a non-Windows library*/
  #define LIBXMLPP_API
#endif /* GLIBMM_DLL */

#ifndef LIBXMLPP_VISIBILITY_DEFAULT
   #define LIBXMLPP_VISIBILITY_DEFAULT
#endif
#ifndef LIBXMLPP_MEMBER_METHOD
  #define LIBXMLPP_MEMBER_METHOD LIBXMLPP_API
#endif
```

and mark the `xmlpp::exception` class in `libxml++/exceptions/exception.h` like:
```
class LIBXMLPP_VISIBILITY_DEFAULT exception : public std::exception
{
public:
  LIBXMLPP_MEMBER_METHOD
  explicit exception(const ustring& message);
  LIBXMLPP_MEMBER_METHOD ~exception() noexcept override;

  LIBXMLPP_MEMBER_METHOD const char* what() const noexcept override;

  LIBXMLPP_MEMBER_METHOD virtual void raise() const;
  LIBXMLPP_MEMBER_METHOD virtual exception* clone() const;

private:
  ustring message_;
};
```

With blessings, and cheers!

[1]: https://gitlab.gnome.org/GNOME/glibmm/-/issues/90

**Comment by fanc999** [comment]
Hi,

I opened 2 PR's:

*  Try to fix linking on MinGW (PR #46)
*  Eliminate warnings regarding `friend class` or `friend struct` nembers (PR #47)

Please let me know how things went with them.

With blessings, thank you!

**Comment by kjellahl** [comment]
> The resource compiler ...
> It probably needs to be defined in your Meson cross-compilation file

That was the only missing part. The resource compiler had been installed,
when I installed mingw-w64.

I've installed `wine64` now, so I can run the example programs.
